docs/odcs.rst

```diff
@@ -0,0 +1,123 @@
+ODCS Integration
+================
+
+When a container build is started, the user usually provides the URL to a yum
+repo that contains the packages to be installed when building the container
+image. Usually, this is a puddle containing signed packages.
+
+In the case a yum repo URL is not provided, and Koji integration is enabled,
+OSBS falls back to the yum repo associated with the build tag for the Koji build
+target (required parameter for Koji container builds). Packages in this yum repo
+are not always signed. This is not ideal as it’s trivial to produce a container
+image that should never be shipped.
+
+Another problem with the current state is that to produce an image with signed
+packages, a parameter is needed (yum repo URL). To complicate matters, this yum
+repo URL will likely be different for future builds. This poses a huge problem
+for OSBS autorebuilds. In this case, OSBS will automatically rebuild layered
+images upon detected changes to parent image. It is necessary that these
+rebuilds pick up the latest, most relevant, signed packages.
+
+The On Demand Compose Service (`odcs`_) project has been created to address
+current limitations in obtaining a yum repo with signed packages. The general
+idea is to not require the user to provide a yum repo url when building
+container images.  Instead, OSBS will use ODCS to generate a yum repo with the
+relevant signed packages based on user configuration.
+
+API Changes
+-----------
+
+The following parameters will be used by ``buildContainer`` task defined by
+``koji-containerbuild`` plugin:
+
+``odcs_signed <bool, default True>`` When set to False, OSBS will provide an
```

Comment:
> I think "ODCS" here is a detail and shouldn't in the parameter name.

It's a detail of the implementation, true. But I don't think we can get away with something less specific without causing confusion. If this parameter was called simply `signed`, it's ambiguous that it only applies to when requesting a new compose from ODCS. Perhaps when ODCS has ubiquitous adoption we may want to revisit.



> The behavior with respect to sigkeys in container.yaml should be specified. Is it?
> 
> If False, sigkeys in container.yaml is ignore
> If True, a non-empty list of sigkeys is required, and one of them must be found.

`sigkeys` in `container.yaml` is optional. If not defined, the OSBS environment will define the default value. But the user has the option to change it (use beta sigkey for example). If the user does decide to use different sigkeys, the list must contain at least one item. (maybe this needs to be written out more explicitly in design doc?) 

To make ODCS not care about package signatures, an empty list of `sigkeys` must be provided. So when `odcs_signed` is set to `False`, an empty list of signkeys is always passed when requesting a compose from ODCS. Otherwise, it uses either what's in container.yaml or the default value in OSBS environment.



> Does it really make sense to say "definitely give me a package from a signed compose, but I don't care if it's provisionally signed or finally signed".

If the intent is "provisional", why should the compose fail if one of the packages (could be dependency) is signed with the "final" key?



> My expectation is that the person (or service) running the containerBuild task should provide an intent ("unsigned", "provisional", "public")

I like this approach. Good call on pointing out multiple repos with key IDs checked in. Providing the intent should allow OSBS to better manage which keys are used. I think a mapping, as you suggested, makes perfect sense. For the "unsigned" intent, it'll essentially map to an empty list of sigkeys.

Reply:
You are right that the "provisional" case provides a case where a single intent can map to a list of signatures - it's fine to get "final" packages in that case. It's the other way that doesn't work.